cfgmgr/sflowmgr.cpp

```diff
@@ -0,0 +1,376 @@
+    {
+        FieldValueTuple fv1("admin_state", local_info.admin);
+        fvs.push_back(fv1);
+    }
+
+    FieldValueTuple fv2("sample_rate", local_info.rate);
+    fvs.push_back(fv2);
+}
+
+void SflowMgr::sflowCheckAndFillValues(string alias, vector<FieldValueTuple> &fvs)
+{
+    string rate;
+    bool admin_present = false;
+    bool rate_present = false;
+
+    for (auto i : fvs)
+    {
+        if (fvField(i) == "sample_rate")
+        {
+            rate_present = true;
+            m_sflowPortConfMap[alias].rate = fvValue(i);
+        }
+        if (fvField(i) == "admin_state")
+        {
+            admin_present = true;
+            m_sflowPortConfMap[alias].admin = fvValue(i);
+        }
+    }
+
+    if (!rate_present)
+    {
+        if (m_sflowPortConfMap[alias].rate == "")
+        {
+            string speed = m_sflowPortConfMap[alias].speed;
+
+            if (speed != SFLOW_ERROR_SPEED_STR)
+            {
+                rate = sflowSpeedRateInitMap[speed];
+            }
+            else
+            {
+                rate = SFLOW_ERROR_SPEED_STR;
+            }
+            m_sflowPortConfMap[alias].rate = rate;
+        }
+        FieldValueTuple fv("sample_rate", m_sflowPortConfMap[alias].rate);
+        fvs.push_back(fv);
+    }
+
+    if (!admin_present)
+    {
+        if (m_sflowPortConfMap[alias].admin == "")
+        {
+            /* By default admin state is enable if not set explicitely */
+            m_sflowPortConfMap[alias].admin = "up";
+        }
+        FieldValueTuple fv("admin_state", m_sflowPortConfMap[alias].admin);
+        fvs.push_back(fv);
+    }
+}
+
+void SflowMgr::doTask(Consumer &consumer)
+{
+    SWSS_LOG_ENTER();
+
+    auto table = consumer.getTableName();
+
+    if (table == CFG_PORT_TABLE_NAME)
+    {
+        sflowUpdatePortInfo(consumer);
+        return;
+    }
+
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+
+        string key = kfvKey(t);
+        string op = kfvOp(t);
+        auto values = kfvFieldsValues(t);
+
+        if (op == SET_COMMAND)
+        {
+            if (table == CFG_SFLOW_TABLE_NAME)
+            {
+                for (auto i : values)
+                {
+                    if (fvField(i) == "admin_state")
+                    {
+                        bool enable = false;
+                        if (fvValue(i) == "up")
+                        {
+                            enable = true;
+                        }
+                        if (enable == m_gEnable)
+                        {
+                            break;
+                        }
+                        m_gEnable = enable;
+                        sflowHandleService(enable);
+                        if (m_intfAllConf)
+                        {
+                            sflowHandleSessionAll(enable);
+                        }
+                        sflowHandleSessionLocal(enable);
```

Comment:
Typical practice for `admin_state` toggle is just  to propagate the field-value tuple change to APPL_DB. Here, when looking into these two function internals, `admin_state`: `down` triggers deleting the entire key, say `Ethernet1`, in APPL_DB `SFLOW_SESSION` table instead of an update to field `admin_state` under the key. This is consistent with the [HLD](https://github.com/Azure/SONiC/blob/master/doc/sflow/sflow_hld.md). But quite curious about the design thought behind it.

Reply:
@wendani : Sflow has the following commands
1) config sflow enable ------------> Enable/disable feature as a whole. When this is disabled, it means all sflow configurations need to be removed (Feature level disable).
2) config sflow interface all enable/disable ----> Global level sflow command to control the enabling/disabling sflow globally. By default when sflow feature is enabled, all ports will have flow enabled. The main usage of this command is in 'disable' configuration. When users need to enable sflow on only one interface or very few interfaces, without global interface disable command, the user has to configure individual interface disable on the rest of the interfaces except for the interfaces of interest. Thus to avoid this, users can have sflow global interface all disable and then do individual sflow enable on interested interfaces which would override the global setting.
3) config sflow interface Ethernetx enable/disable - Local interface level command which would override global interface all command when configured on specific interfaces